### ISSUE TITLE
Accessing obviously non-existent class attribute not seen as an error

### ISSUE THREAD

**Issue by DanielSank** [python/mypy#12122]
Warning: I didn't ask about this on gitter because the required permissions seemed scary.

**Bug Report**

mypy fails to recognize an obvious type error. I have checked the help pages and I *think* my case (shown below) is not one of the known gotchas.

**To Reproduce**

Run `mypy` against this file:
```python
from typing import Any, cast, Generic, Type, TypeVar

class Foo:
    x: int
    y: str
    def __init__(self, x: int, y: str):
        self.x = x
        self.y = y

T = TypeVar('T')

class Serdes(Generic[T]):
    """Helps serialize and deserialize objects to string."""
    klass: Type[T]

    def __init__(self, klass: Type[T]):
        self.klass = klass

    @staticmethod
    def to_str(obj: T) -> str:
        print(obj)  # For lulz
        return "123"  # Fake serialized version of obj

    def from_str(self, s: str) -> T:
        obj = cast(T, _deserialize(s))
        return obj

def _deserialize(s: str) -> Any:  # Is this Any the problem?
    """Fake deserialization that always returns Foo(4, 'foo')"""
    print(s)  # for lulz
    return Foo(x=4, y='foo')

def example():
    foo = Foo(x=4, y='foo')
    serdes = Serdes(klass=Foo)
    foo_serialized = serdes.to_str(foo)
    foo_deserialized = serdes.from_str(foo_serialized)

    foo_deserialized.z
    # mypy doesn't flag this as an error. Interestingly, my in-editor linter
    # (pyright) does flag it as an error.
```

**Expected Behavior**

I expect the final line of the code to be identified as a type error.

**Actual Behavior**

Mypy says all is well.

**Your Environment**

- Mypy version used: `0.812`
- Mypy command-line flags: None
- Mypy configuration options from `mypy.ini` (and other config files): This is a good question... where is this file when using a virtualenv?
- Python version used: `3.8.8`
- Operating system and version: Debian variant

**Comment by JelleZijlstra** [comment]
Can you use `reveal_type()` to figure out where mypy is going wrong? For example, what type does mypy think `foo_deserialized` has? What about the other variables in the function?

**Comment by DanielSank** [comment]
TIL about `reveal_type()`.

Here's the output when I use `reveal_type()` on `foo_deserialized`
```
typecheck.py:38: note: Revealed type is 'Any'
typecheck.py:38: note: 'reveal_type' always outputs 'Any' in unchecked functions
```

**Comment by JelleZijlstra** [comment]
Ah, I missed that! You hit https://mypy.readthedocs.io/en/stable/common_issues.html#no-errors-reported-for-obviously-wrong-code

**Comment by DanielSank** [comment]
Thanks. I read those notes before posting here. Why didn't the cast and/or the `from_string` annotations fix this up? When doing IO we always have to typecast somewhere. How do we inform mypy of what's going on?

**Comment by JelleZijlstra** [comment]
The primary problem is that `def example` is unannotated.

**Comment by DanielSank** [comment]
OH.

Thanks.

It's fine now.
